### PR TITLE
Feature/feat 2 authentication flow

### PR DESCRIPTION
**Pull request by Peregrinus0304**

### CODE REVIEW

File: .swiftlint.yml [hunk 1]
```diff
@@ -1 +1 @@
...
    warning: 300
    error: 500
file_length:
    warning: 1000
    error: 1500
    ignore_comment_only_lines: true
cyclomatic_complexity:
    warning: 15
    error: 25
reporter: "xcode"
```
Comment (dzoanb) [inline .swiftlint.yml]: Are you editing this in TextEdit? Please don't, it messes up the end end of line sign as you can see. Also, may be interesting for you to challange yourself with 120 line lenghts, like we have in the project 😅

File: FantasyUniverse.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -561,7 +593,7 @@
 			);
 			runOnlyForDeploymentPostprocessing = 0;
 			shellPath = /bin/sh;
-			shellScript = "if [[ -f \"${PROJECT_DIR}/SwiftGen/bin/swiftgen\" ]]; then\n\necho $SRCROOT\n\n  rm \"$SRCROOT/FantasyUniverse/Generated/Strings.swift\"\n\n  rm \"$SRCROOT/FantasyUniverse/Generated/Assets.swift\"\n\n  \"${PROJECT_DIR}/SwiftGen/bin/swiftgen\" config run\n\nelse\n\n  echo \"warning: SwiftGen is not installed. Run 'pod install --repo-update' to install it.\"\n\nfi\n";
+			shellScript = "if [ $ACTION != \"indexbuild\" ]; then\nif [[ -f \"${PROJECT_DIR}/SwiftGen/bin/swiftgen\" ]]; then\n\necho $SRCROOT\n\n  rm \"$SRCROOT/FantasyUniverse/Generated/Strings.swift\"\n\n  rm \"$SRCROOT/FantasyUniverse/Generated/Assets.swift\"\n\n  \"${PROJECT_DIR}/SwiftGen/bin/swiftgen\" config run\n\nelse\n\n  echo \"warning: SwiftGen is not installed. Run 'pod install --repo-update' to install it.\"\n\nfi\nfi\n\n";
```
Comment (dzoanb) [inline FantasyUniverse.xcodeproj/project.pbxproj:602]: That's interesting, why is this change needed?

File: FantasyUniverse/Services/Authentication/FirebaseAuthService.swift [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+//  FantasyUniverse
+//
+//  Created by TarasPeregrinus on 17.05.2022.
+//
+
+import Foundation
+import FirebaseAuth
+import Firebase
+
+class FirebaseAuthService {
```
Comment (dzoanb) [inline FantasyUniverse/Services/Authentication/FirebaseAuthService.swift]: Maybe consider a reactive interface using Combine or trying out the Async/Await? It would incentivise you to handle errors more gracefully - it's best to have some errors returned and adequate errors presented to the user, so I wouldn't hide them all under a boolean.

File: FantasyUniverse/Services/Authentication/FirebaseAuthService.swift [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            let userData = ["firstName": name,
+                            "lastName": surname]
+            ref.child("users").child(Auth.auth().currentUser!.uid).setValue(userData)
+        } else {
+            debugPrint("Error: Trying to store user info for signed out user")
+        }
+    }
+    
+    func signIn(email: String, pass: String, completionBlock: @escaping (_ success: Bool) -> Void) {
+        Auth.auth().signIn(withEmail: email, password: pass) { (result, error) in
```
Comment (dzoanb) [inline FantasyUniverse/Services/Authentication/FirebaseAuthService.swift]: Isn't there any token returned or is it handled by Firebase internally?

File: FantasyUniverse/Services/AuthenticationValidator.swift [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    let isPasswordValid: (String?) -> (Bool) = { (password: String?) -> (Bool) in
+        let passwordPred = NSPredicate(format: "SELF MATCHES %@ ", "^(?=.*[a-z])(?=.*[0-9])(?=.*[A-Z])(?=.*[!@#$&*]).{8}$")
+        return passwordPred.evaluate(with: password)
+    }
+    
+    let isNameValid: (String?) -> (Bool) = { (name: String?) -> (Bool) in
+        let namePred = NSPredicate(format: "SELF MATCHES %@ ", "^[A-Za-z]{2,32}$")
+        return namePred.evaluate(with: name)
+    }
+}
```
Comment (dzoanb) [inline FantasyUniverse/Services/AuthenticationValidator.swift]: Sexy ✨ But please add tests 😄

File: FantasyUniverse/Views/App views/CredentialsInputView.swift [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+//  Created by TarasPeregrinus on 14.05.2022.
+//
+
+import SwiftUI
+
+struct CredentialsInputView: View {
+    
+    var labelText: String
+    var textFieldPlaceholder: String
+    var isSecure: Bool
```
Comment (dzoanb) [inline FantasyUniverse/Views/App views/CredentialsInputView.swift]: These could be constants, right?

File: FantasyUniverse/Views/App views/AppTextFieldStyles.swift [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+      .font(.title.weight(.semibold))
+      .padding(.vertical, 12)
+      .padding(.horizontal, 16)
+      .background(
+        LinearGradient(gradient: Gradient(colors: [Color(Asset.Colors.appPurple.color), .blue]), startPoint: .top, endPoint: .bottom)
+          .edgesIgnoringSafeArea(.all))
+      .frame(height: 50)
+      .cornerRadius(15)
+  }
+}
```
Comment (dzoanb) [inline FantasyUniverse/Views/App views/AppTextFieldStyles.swift]: Most of the project has 4-wide indentation, please fix.

File: FantasyUniverse/Views/SignUpView.swift [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+import SwiftUI
+
+struct SignUpView: View {
+    @ObservedObject var viewModel = SignUpViewModel()
+    
+    @State private var firstNameFieldValue = ""
+    @State private var lastNameFieldValue = ""
+    @State private var loginFieldValue = ""
+    @State private var passwordFieldValue = ""
```
Comment (dzoanb) [inline FantasyUniverse/Views/SignUpView.swift]: If you're going for MVVM pattern then these should be published properties in the view model and check credentials should react on the changes in view model so that logic is not leaking into views.

**Review by dzoanb — APPROVED** [review]

File: .swiftlint.yml (hunk 1 above)
Comment (Peregrinus0304) [reply .swiftlint.yml]: Ok, use Sublime Text e.g. instead ?

File: FantasyUniverse.xcodeproj/project.pbxproj (hunk 2 above)
Comment (Peregrinus0304) [reply FantasyUniverse.xcodeproj/project.pbxproj:602]: Hm.. as far as I remember it fixed some swiftgen code generating issues.

File: FantasyUniverse/Services/Authentication/FirebaseAuthService.swift (hunk 4 above)
Comment (Peregrinus0304) [reply FantasyUniverse/Services/Authentication/FirebaseAuthService.swift]: Yeah, we can retrieve token from result like result?.user.refreshToken or get idToken like firebase.auth().currentUser.getIdToken, but i will handle it later on

File: FantasyUniverse/Services/Auth/FirebaseAuthService.swift [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+                                    .updateChildValues(values) { error, _ in
+                                        if let err = error {
+                                            promise(.failure(err))
+                                        } else {
+                                            promise(.success(()))
+                                        }
+                                    }
+                                
+                            } else {
+                                promise(.failure(NSError(domain: "Invalid User Id", code: 0, userInfo: nil)))
```
Comment (dzoanb) [inline FantasyUniverse/Services/Auth/FirebaseAuthService.swift:99]: Make and enum like "FirebaseAuthError" with cases for errors like this `case invalidUserID`.
